### PR TITLE
Metadata Uploader

### LINKED ISSUE #796: Add uploading comma- or tab-delimited metadata files
## Describe your idea for a new feature
Metadata upload (**Project** -> **Line List** -> **Upload Metadata**) should accept delimited text files like CSV or TSV as an acceptable format.

## Describe how you would access this feature?
Currently, uploading metadata can only be done via MS Excel files. You would be able to upload comma- or tab-delimited text files in the same manner as `.xlsx` files. 
 
## Additional information
Requiring MS Excel as a processing step if the metadata doesn't already live in Excel adds complexity and raises compatibility issues. And as we know, [Excel ](http://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1044-7) [has](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-5-80) [problems](http://www.eusprig.org/horror-stories.htm).

### PR DESCRIPTION
**Pull request by ksierks**
## Description of changes

image
image
image
image
image

## Related issue
Fixes #796 & #84.

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

* [x] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.
* [x] Tests added (or description of how to test) for any new features.
* [x] User documentation updated for UI or technical changes.

### CODE REVIEW

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/dto/SavedMetadataErrorResponse.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package ca.corefacility.bioinformatics.irida.ria.web.projects.dto;
+
+import ca.corefacility.bioinformatics.irida.ria.utilities.SampleMetadataStorage;
+import ca.corefacility.bioinformatics.irida.ria.web.ajax.dto.ajax.AjaxResponse;
+
+/**
+ * Returns the SampleMetadataStorage on error.
```
Comment (joshsadam) [inline src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/dto/SavedMetadataErrorResponse.java:7]: Isn't this the same as `SavedMetadataException`?  They seem to do the exact same thing so just have one.

File: src/main/java/ca/corefacility/bioinformatics/irida/service/impl/ProjectServiceImpl.java [hunk 2]
```diff
@@ -266,7 +266,8 @@ public void removeUserFromProject(Project project, User user) throws ProjectWith
 	public void removeUserGroupFromProject(Project project, UserGroup userGroup) throws ProjectWithoutOwnerException {
 		final UserGroupProjectJoin j = ugpjRepository.findByProjectAndUserGroup(project, userGroup);
 		if (!allowRoleChange(project, j.getProjectRole())) {
-			throw new ProjectWithoutOwnerException("Removing this user group would leave the project without an owner.");
```
Comment (joshsadam) [inline src/main/java/ca/corefacility/bioinformatics/irida/service/impl/ProjectServiceImpl.java:269]: There are a lot of random formatting changes to this file.  Make sure IntelliJ is set to format only VCS changes.  Don't worry about fixing this.

File: src/main/resources/i18n/messages.properties [hunk 3]
```diff
@@ -1961,40 +1965,44 @@ ViewAnnouncement.read=Read
...
-metadata.results.missing.none-title=Success!
-metadata.results.missing.none-text=All rows mapped to samples in this project, nothing to worry about.
-
-metadata.results.save.found-error=Since none of the metadata could match existing samples, nothing could be save.
-metadata.results.save.sample-not-found=Sample could not be found: {0}
-metadata.results.save.success={0} samples have had metadata added to them.
+
+server.metadataimport.results.save.success.single-updated={0} sample has been updated
+server.metadataimport.results.save.success.multiple-updated={0} samples were updated
+server.metadataimport.results.save.success.single-created=\ and {0} new sample was created.
```
Comment (joshsadam) [inline src/main/resources/i18n/messages.properties:1971]: I did not notice this before, but this was a really smart way to handle this!

File: src/main/webapp/pages/projects/project_samples.html [hunk 4]
```diff
@@ -4,47 +4,53 @@
   xmlns:layout="http://www.ultraq.net.nz/thymeleaf/layout"
```
Comment (joshsadam) [inline src/main/webapp/pages/projects/project_samples.html:4]: Why so many formatting changes?  You are going to create a huge headache for me when I merge this with my share samples branch.

File: src/main/webapp/pages/projects/project_samples_metadata_upload.html [hunk 5]
```diff
@@ -1,132 +1,18 @@
 <!DOCTYPE html>
 <html
-      xmlns:th="http://www.thymeleaf.org"
-      xmlns:layout="http://www.ultraq.net.nz/thymeleaf/layout"
-      data-layout-decorate="~{projects/_base}">
```
Comment (joshsadam) [inline src/main/webapp/pages/projects/project_samples_metadata_upload.html:5]: Inside your root tag on this page, can you add a comment about the path to the base javascript file.

File: src/main/webapp/resources/js/apis/metadata/sample-utils.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Utility file for samples
+ */
+
+export const sampleNameRegex = new RegExp("^[A-Za-z0-9-_]{3,}$");
+
```
Comment (joshsadam) [inline src/main/webapp/resources/js/apis/metadata/sample-utils.js:6]: Method comment

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportMapHeaders.jsx [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+          </BlockRadioInput>
+        ))}
+      </Radio.Group>
+      <div style={{ display: "flex" }}>
+        <Button
+          className="t-metadata-uploader-file-button"
+          icon={<IconArrowLeft />}
+          onClick={() => history.goBack()}
+        >
+          {" "}
```
Comment (joshsadam) [inline src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportMapHeaders.jsx:65]: Get rid of this empty space.

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportWizard.jsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+ */
+export function SampleMetadataImportWizard({
+  currentStep,
+  currentStatus,
+  children,
+}) {
+  const { projectId } = useParams();
+
+  return (
+    <Space direction="vertical" size="large" style={{ width: `100%` }}>
```
Comment (joshsadam) [inline src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportWizard.jsx:23]: `size="large"` is actually not an option only small, I think large is the default.

**Review by joshsadam — CHANGES_REQUESTED** [review]
Just a couple of changes and I think this might be good to go.

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportWizard.jsx (hunk 8 above)
Comment (ksierks) [reply src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportWizard.jsx:23]: image

The ant design documentation says 'small' is the default and 'large' is an option.

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/ProjectsController.java [hunk 9]
```diff
@@ -91,8 +90,8 @@
 
 	@Autowired
 	public ProjectsController(ProjectService projectService, SampleService sampleService, UserService userService,
-			ProjectControllerUtils projectControllerUtils, TaxonomyService taxonomyService,
-			UICartService cartService, UpdateSamplePermission updateSamplePermission, MessageSource messageSource) {
+			ProjectControllerUtils projectControllerUtils, TaxonomyService taxonomyService, UICartService cartService,
```
[the remark this answers is not in the document]
Comment (ksierks) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/ProjectsController.java]: Please see 461911e85b4a8663b9b70f30065666ca1d70fec5.

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportMapHeaders.jsx (hunk 7 above)
Comment (ksierks) [reply src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportMapHeaders.jsx]: Please see 34cb910e3d0ca2b2b205e0854c2b600b3cd98a7a.

File: src/main/webapp/resources/js/apis/metadata/sample-utils.js (hunk 6 above)
Comment (ksierks) [reply src/main/webapp/resources/js/apis/metadata/sample-utils.js:6]: Please see 891a43efa0566e9386376b942e9e528363aceef8.

File: src/main/webapp/pages/projects/project_samples_metadata_upload.html (hunk 5 above)
Comment (ksierks) [reply src/main/webapp/pages/projects/project_samples_metadata_upload.html:5]: Please see 0084bfc44642de61afea58b83d4d6591465de280.

File: src/main/resources/i18n/messages.properties (hunk 3 above)
Comment (ksierks) [reply src/main/resources/i18n/messages.properties:1971]: Thanks.

File: src/main/webapp/pages/projects/project_samples.html (hunk 4 above)
Comment (ksierks) [reply src/main/webapp/pages/projects/project_samples.html]: Sorry, editor troubles... please see e50b2deee76325179cb51cc1d63b5119a5b12610.

File: src/main/java/ca/corefacility/bioinformatics/irida/service/impl/ProjectServiceImpl.java (hunk 2 above)
Comment (ksierks) [reply src/main/java/ca/corefacility/bioinformatics/irida/service/impl/ProjectServiceImpl.java:269]: Okies. Thank you for the screenshot.

Comment (joshsadam) [reply src/main/java/ca/corefacility/bioinformatics/irida/service/impl/ProjectServiceImpl.java:269]: :) You're welcome.

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportWizard.jsx (hunk 8 above)
Comment (joshsadam) [reply src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportWizard.jsx:23]: :facepalm:

**Review by joshsadam — APPROVED** [review]

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/dto/SavedMetadataErrorResponse.java (hunk 1 above)
Comment (joshsadam) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/dto/SavedMetadataErrorResponse.java:7]: We will keep them as they have slightly different useage.

**Pull request merged**